### PR TITLE
Error handling

### PR DESCRIPTION
**Pull request by Danielwsx64**
This pull request close the issue-2 https://github.com/brainn-co/xcribe/issues/2

It closes the issue-33 https://github.com/brainn-co/xcribe/issues/33 too

We adopt some strategies to improve error handling,

A struct `Request.Error` to wrapping errors. It is used as return by `Conn.Parser` when it cant parse a given `Plug.Conn`. Next step is create a `Request.Validator` to validate all given requests before send them to document module and wrap the errors in that struct (This will be done in a future PR).

`Xcribe.Document.document/2` macro append to the structs `Resquest` and `Request.Error` a meta information about the file and the line where it was called. This information will be used later to help user locate the macro call when happen some error.

`Xcribe.DocException` was created to wrapping  exceptions at documenting time. `rescue` statements were placed in strategic points to catch any exception and merge it with the request information. It allow us identify where the request how raised the exception was originally given to `Xcribe.Document.document/2`.

A new function `Xcribe.Config.check_configurations/0` was created to validate sensitive configuration.

So then `Xcribe.Formatter` centralize the error handling. First of all it checks Xcribe configurations. Next it checks if some parse error from `Xcribe.ConnParser` as added to the` Recorder` (I decided to check it in Formatter to not interfere in users tests flow, so after all tests end we check for errors). If all is ok the requests are ordered by their paths and sent to documentation module. If some exception happen the `Formatter` will cath it. If any one of this errors happen (configuration, parsing or exception while documenting) the `Formatter` handle the errors and pass them to the new module` Xcribe.CLI.Output` and it prints a friendly message about the errors.

I copied the same layout the Credo library uses :P

xcribe-exception-output
xcribe-configuration-error-output
xcribe-parsing-error-output

Some future improvements we'll need
- ApiBlueprint is messy and now we have a lot of `rescue-raise` in it. Like we made with `Swagger` whe need decrease the Blueprint Formatter public API. This way we can decrease the side effect points.
- Implement the https://github.com/brainn-co/xcribe/issues/19 proposal 
- I would like to refactoring the new `Xcribe.CLI.Output`. I'm think about have some `mix tasks`, and would be useful if the `Output` module was more generally.
- Some exceptions I created to configuration are not needed any more.
- Was not a god idea keep the old configuration keys.

### CODE REVIEW

File: test/lib/cli/output_test.exs [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+        __meta__: %{
+          call: %{
+            description: "test name",
+            file: File.cwd!() <> "/test/lib/cli/output_test.exs",
+            line: 13
+          }
+        }
+      }
+
+      # |> document(as: "awesome route")
```
Comment (brosquinha) [inline test/lib/cli/output_test.exs:27]: :eyes:

File: test/lib/cli/output_test.exs [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+  alias Xcribe.CLI.Output
+  alias Xcribe.DocException
+  alias Xcribe.Request.Error
+
+  import ExUnit.CaptureIO
+
+  describe "print_request_errors/1" do
+    test "parsing errors" do
+      # |> document(as: "some cool description")
```
Comment (brosquinha) [inline test/lib/cli/output_test.exs:13]: :eyes:

File: lib/cli/output.ex [hunk 3]
```diff
@@ -0,0 +1,111 @@
+defmodule Xcribe.CLI.Output do
```
Comment (thiamsantos) [inline lib/cli/output.ex:1]: Nice :heart_eyes:

File: lib/config.ex [hunk 4]
```diff
@@ -52,17 +55,22 @@ defmodule Xcribe.Config do
...
 
       config :xcribe, [
         information_source: YourApp.YouModuleInformation
       ]
   """
-  def xcribe_information_source do
+  def xcribe_information_source, do: get_xcribe_config(:information_source)
+
+  @doc """
+  Return the iformation module with API information
```
Comment (thiamsantos) [inline lib/config.ex:69]:
```suggestion
  Return the information module with API information
```

File: lib/config.ex [hunk 5]
```diff
@@ -52,17 +55,22 @@ defmodule Xcribe.Config do
   def active?, do: !is_nil(System.get_env(env_var_name()))
 
   @doc """
-  Return the iformation module with API information (`Xcribe.Information`).
-
-  If information source is not given an `Xcribe.MissingInformationSource` exception will raise.
+  Return the iformation module with API information
```
Comment (thiamsantos) [inline lib/config.ex:58]:
```suggestion
  Return the information module with API information
```

File: lib/config.ex [hunk 6]
```diff
@@ -83,18 +91,79 @@ defmodule Xcribe.Config do
...
+  @doc """
+  Return ok if given configurations are valid.
+
+  If same invalid config was set an tuple with a list of erros will be returned.
+  """
+  def check_configurations,
+    do: Enum.reduce([:format, :information_source, :json_library], :ok, &validate_config/2)
+
+  @format_message "Xcribe doe't support the configured documentaion format"
+  @format_instructions "Xcribe supports Swagger and Blueprint, configure as: `config :xcribe, [format: :swagger]`"
```
Comment (thiamsantos) [inline lib/config.ex:103]:
```suggestion
  @format_instructions "Xcribe supports Swagger and Blueprint, configure as: `config :xcribe, format: :swagger`"
```

File: lib/config.ex [hunk 7]
```diff
@@ -83,18 +91,79 @@ defmodule Xcribe.Config do
...
 
+  @doc """
+  Return ok if given configurations are valid.
+
+  If same invalid config was set an tuple with a list of erros will be returned.
+  """
+  def check_configurations,
+    do: Enum.reduce([:format, :information_source, :json_library], :ok, &validate_config/2)
+
+  @format_message "Xcribe doe't support the configured documentaion format"
```
Comment (thiamsantos) [inline lib/config.ex:102]:
```suggestion
  @format_message "Xcribe doesn't support the configured documentaion format"
```

File: lib/config.ex [hunk 8]
```diff
@@ -83,18 +91,79 @@ defmodule Xcribe.Config do
...
+        :information_source,
+        module,
+        @info_source_message,
+        @info_source_instructions
+      )
+    end
+  end
+
+  @json_lib_message "The configured json library doesn't implement the needed functions"
+  @json_lib_instructions "Try configure Xcribe with Jason or Poison `config :xcribe, [json_library: Jason]`"
```
Comment (thiamsantos) [inline lib/config.ex:133]:
```suggestion
  @json_lib_instructions "Try configure Xcribe with Jason or Poison `config :xcribe, json_library: Jason`"
```

File: lib/conn_parser.ex [hunk 9]
```diff
@@ -42,6 +52,13 @@ defmodule Xcribe.ConnParser do
     |> router_module()
     |> apply(:__match_route__, [method, decode_uri(path), host])
     |> extract_route_info()
+  rescue
+    _ ->
+      Map.put(
+        @error_struct,
```
Comment (thiamsantos) [inline lib/conn_parser.ex:58]:
```suggestion
        %{@error_struct | message: "An invalid Plug.Conn was given or maybe an invalid Router"}
```
:nail_care:

File: test/lib/cli/output_test.exs (hunk 1 above)
Comment (Danielwsx64) [reply test/lib/cli/output_test.exs:27]: I need a line with to test the output. Test if we are able to point the document macro.

Comment (brosquinha) [reply test/lib/cli/output_test.exs:27]: Ah, I see... maybe something to make this clearer? :thinking: 

And what about line 13? What is that comment doing?

Comment (Danielwsx64) [reply test/lib/cli/output_test.exs:27]: Do you mean a comment to explain the comments? 

(the line 13 has the same purpose)

**Pull request merged**